New component: Python Script

balloob opened the pull request:
## Description:
This is an experimental component. Hacked together after a discussion with @bbangert about adding Lambda support to Hass.

So what is it? This is a component to allow you to write Python scripts that are exposed as services in Home Assistant. Each Python file created in the `<config>/python_scripts/` folder will be exposed as a service. The content is not cached so you can easily develop: edit file, save changes, call service. The scripts are run in a sandboxed environment with access to the `hass` object, the service call data as `data` and a logger as `logger`.

The idea is that writing Python automation components is often overlooked as an option. This will make it easier to use our existing automation component for the trigger and then use a Python script to handle the execution.

So, to try this out:

 - Add to `configuration.yaml`: `python_script:`
 - Create folder `<config>/python_scripts`
 - Create a file `hello_world.py` and give it this content:

```python
name = data.get('name', 'world')
logger.info("Hello {}".format(name))
hass.bus.fire(name, { "wow": "from a Python script!" })
```

 - Start Home Assistant
 - Call service `python_script/hello_world`

![image](https://user-images.githubusercontent.com/1444314/26915704-f97a6a52-4bdb-11e7-8e21-d6d79025976b.png)

And look at your console:

![image](https://user-images.githubusercontent.com/1444314/26915740-14dd154c-4bdc-11e7-9b1c-2591e4b958c7.png)

To do:
 - [ ] think about if this is a good idea
 - [ ] tests
 - [ ] docs

Feedback welcome if this would be a good idea or not.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#<home-assistant.github.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml
python_script:
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code does not interact with devices:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

balloob committed:
Add initial version

balloob committed:
Fix requirements

balloob committed:
Prefer logging over printing

balloob commented:
@MartinHjelmare good call. `logger` will now be a `logging.Logger` instance. When printing, it will instead print a warning to use the logger.

balloob committed:
Set executor thread name on >Py36 only

balloob committed:
Add tests

balloob committed:
Lint

balloob committed:
Add restrictedpython to test dependencies

MartinHjelmare commented on tests/components/test_python_script.py:
```diff
@@ -0,0 +1,123 @@
+from homeassistant.components.python_script import execute
+
+
+@asyncio.coroutine
```
How is this picked up by our tester? Is it builtin in pytest?

pvizeli replied:
yes

pvizeli committed:
Create python_script.py

From doc:
```
However, an empty dict ({}) is treated as is. If you want to specify a list that can contain anything, specify it as dict:
>>> schema = Schema({}, extra=ALLOW_EXTRA)  # don't do this
>>> try:
...   schema({'extra': 1})
...   raise AssertionError('MultipleInvalid not raised')
... except MultipleInvalid as e:
...   exc = e
>>> str(exc) == "not a valid value"
True
>>> schema({})
{}
>>> schema = Schema(dict)  # do this instead
>>> schema({})
{}
>>> schema({'extra': 1})
{'extra': 1}

```

pvizeli reviewed (approved):
Yeah :dancer:

CoalaJoe commented:
Does this allow inline(in yaml) python?
Otherwise, you could just make the script and add it as a CLI command. With a shebang header, you don't even have to add the python executable.

postlund commented:
I know this is closed now, but I have a small question... How safe is this component regarding blocking of event loop if the script performs something that is time demanding/blocking? To do more complex operations, I assume I must add a job/coro to the event loop thus allowing me to yield from that?

@CoalaJoe It does not. It gives the benefit that you do not have to instantiate and use the websocket client library when communicating with Home Assistant since the script is run inside the same process as Home Assistant. If the script does not use any data from Home Assistant (like sensor values), then I see less value and a regular shell command would work too (you don't have to spawn a new process using this method though).

CoalaJoe commented:
@postlund Ah. Thanks for clearing that out.

balloob commented on homeassistant/components/python_script.py:
```diff
@@ -11,7 +11,7 @@
 
 CONFIG_SCHEMA = vol.Schema({
-    DOMAIN: vol.Schema({})
+    DOMAIN: vol.Schema(dict)
```
Actually, I did want to validate against an empty dictionary. There is no config available.

balloob commented:
@postlund Python scripts are run in executor threads. They cannot yield against the event loop but are able to queue up jobs using `hass.add_job`. I guess you could use a script to define a coroutine, schedule it and then finish, but I haven't tested that.

The next step is going to be to add support for this in Script syntax. That way it will be possible to make the variables from automation triggers directly available inside the script (instead of JSON compatible versions that come in via service calls)
